diesel/src/pg/backend.rs

```diff
@@ -13,6 +13,29 @@ use sql_types::{Oid, TypeMetadata};
 #[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
 pub struct Pg;
 
+#[derive(Debug, Clone)]
+pub struct PgValue {
+    data: Vec<u8>,
```

Comment:
Once https://github.com/rust-lang/rust/issues/48055 lands, we can just have `data: [u8]` as the final field of the struct. Until then, we'll probably need to use `*const [u8]`

Reply:
So as requested by @Thomspoon I've played a bit around with how to write such a struct. As Sean has written above, currently the probably easiest way is to use  `NonNull<[u8]>` as field here.

This would result in the following code:
```rust 
struct PgValue {
    raw_bytes: NonNull<[u8]>,
    oid: u32
}

impl PgValue {
    unsafe fn new(value_ptr: *mut u8, byte_count: usize) -> Self {
        let slice =  from_raw_parts_mut(value_ptr, byte_count);
        Self {
            raw_bytes: NonNull::new(slice).expect("cannot be null"),
        }
    }

    fn as_bytes(&self) -> &[u8] {
        unsafe { self.raw_bytes.as_ref() }
    }

    fn oid(&self) -> u32 {
        self.oid
    }
}
```

The new function should be only called in `PgResult::get`. For constructing `PgValue` from a other `PgValue` from inside a `from_sql` a additional function should be added.